libraries/postgres/src/Database.js

```diff
@@ -0,0 +1,292 @@
+      for (let serviceName of schema.access.serviceNames()) {
+        const username = `${usernamePrefix}_${serviceName.replace(/-/g, '_')}`;
+
+        // calculate the expected privs based on access.yml
+        const tables = schema.access.tables(serviceName);
+        Object.entries(tables).forEach(([table, mode]) => {
+          if (mode === 'read') {
+            expectedPrivs.add(`${username}: SELECT on ${table}`);
+          } else if (mode === 'write') {
+            expectedPrivs.add(`${username}: SELECT on ${table}`);
+            expectedPrivs.add(`${username}: INSERT on ${table}`);
+            expectedPrivs.add(`${username}: UPDATE on ${table}`);
+            expectedPrivs.add(`${username}: DELETE on ${table}`);
+          }
+        });
+      }
+
+      const issues = [];
+      for (let cur of currentPrivs) {
+        if (!expectedPrivs.has(cur)) {
+          issues.push(`unexpected database user grant: ${cur}`);
+        }
+      }
+
+      for (let exp of expectedPrivs) {
+        if (!currentPrivs.has(exp)) {
+          issues.push(`missing database user grant: ${exp}`);
+        }
+      }
+
+      if (issues.length > 0) {
+        throw new Error(`Database privileges are not configured as expected:\n${issues.join('\n')}`);
+      }
+    });
+  }
+
+  async _doUpgrade({version, showProgress, usernamePrefix}) {
+    await this._withClient('admin', async client => {
+      await client.query('begin');
+      if (version.version === 1) {
+        await client.query('create table tcversion as select 0 as version');
+      }
+
+      // check the version and lock it to prevent other things from changing it
+      const res = await client.query('select version from tcversion for update');
+      if (res.rowCount !== 1 || res.rows[0].version !== version.version - 1) {
+        throw Error('Multiple DB upgrades running simultaneously');
+      }
+      showProgress('..running migration script');
+      const migrationScript = version.migrationScript
+        .replace('$db_user_prefix$', usernamePrefix);
+      await client.query(`DO ${dollarQuote(migrationScript)}`);
+      showProgress('..defining methods');
+      for (let [methodName, { args, body, returns}] of Object.entries(version.methods)) {
+        await client.query(`create or replace function
+        "${methodName}"(${args})
+        returns ${returns}
+        as ${dollarQuote(body)}
+        language plpgsql`);
+      }
+      showProgress('..updating version');
+      await client.query('update tcversion set version = $1', [version.version]);
+      showProgress('..committing transaction');
+      await client.query('commit');
+    });
+  }
+
+  /**
+   * Private constructor (use Database.setup and Database.upgrade instead)
+   */
+  constructor({urlsByMode, statementTimeout}) {
+    const makePool = dbUrl => {
+      const pool = new Pool({connectionString: dbUrl});
+      // ignore errors from *idle* connections
+      pool.on('error', client => {});
```

Comment:
Will we really want to throw away 100% of errors from idle connections 100% of the time? I'm wondering if we'll hide some larger issue here by accident.